### DISCUSSION TITLE
Isolating parts of the library using Terser

### DISCUSSION THREAD

**Discussion by belohlavek** [101arrowz/fflate#73] [Q&A] [answered]
Dear @101arrowz 

Hey! I need to isolate parts of this library, preferably keeping the type annotations, but a JS output would be ok too.

For example, I want to "dump" the Inflate and Deflate functions into their own modules.
There are several reasons why I want to do this, including (but not limited to) having an easier time running the code through TypescriptToLua.

Normally I would just do this manually, but having all the exported members in a single file, plus using short names makes the code a bit hard to follow (at least for my _extraction_ purposes).

Can some of this process be automated using Terser? It could be similar to those small bundles which only export a limited set of features. Is it possible to get a clean un-minified output?

Do you have a config or template project to share? I read in another thread that you already have some of these small builds but chose not to provide them due to the (understandably) wide range of possible configurations. I want to generate one of these configurations myself.

Thanks for the great lib!

**Comment by 101arrowz** [comment, marked as answer]
To generate custom builds myself, I usually run the build through `terser`. You can just paste [the CJS build](https://unpkg.com/fflate@0.7.0/lib/browser.cjs) into a text editor, then find+replace all the exports with nothing (e.g. with the regex `exports.*;$` and manually removing `exports.FlateErrorCode`). Now, manually write the exports you want at the bottom (e.g. `exports.unzipSync = unzipSync;`), then plug it through `terser` (since I'm lazy I use [terser-online](https://xem.github.io/terser-online/)). Now you have a minified file that contains the exports, and you can use the TypeScript declaration file [provided with `fflate`](https://unpkg.com/fflate@0.7.0/lib/index.d.ts) to plug it through TypescriptToLua.

Unrelated, but I'm curious why you don't just use bindings to native Zlib written in Lua. They would probably be faster than `fflate` ported to Lua (since `fflate` includes a LOT of micro-optimizations that work best in JavaScript). Also, my own tests on this method don't yield working code because TypescriptToLua isn't actually very accurate: `fflate` assumes that 0 is falsy in a lot of places, [but TypescriptToLua does not](https://typescripttolua.github.io/docs/caveats#boolean-coercion).

**Reply by belohlavek** [reply]
Whoops, good catch on that TypescriptToLua detail, probably saved my quite a bit of time, thank you.
The reason I was considering it is that I can't access or create bindings in the sandboxed environment this lua script would run on (basically an extension for another software). There's [another pure lua library](https://github.com/SafeteeWoW/LibDeflate) but I haven't yet gotten it to inflate correctly with fflate. It may be my lack of lua knowledge but it seems to be focused on strings, while I need a simple array-like structure for numbers. I will probably investigate the project further to understand how it's implemented internally and try to feed the lua table (array) directly.  Just to be clear, this is me incorrectly using that lua library, not a problem with fflate.

As I mentioned before, I can still make use of isolating parts of fflate for other purposes, so thanks for the tips!

**Reply by 101arrowz** [reply]
The library you linked should work; it returns binary strings instead of an array, and seems to accept binary strings as well. In a binary string, basically the character codes in the string correspond to the byte values you want. You can probably convert an array of numbers to a binary string pretty easily in Lua, you're looking for the equivalent of `String.fromCharCode(code)` in JS. To do the reverse, look for the equivalent of `String.charCodeAt`.

`fflate` has binary string manipulation functions as well: `strToU8(binstring, true)` converts binary strings to `Uint8Array` and `strFromU8(someUint8ArrayObject, true)` converts `Uint8Array` to binary strings.
